src/rfc822/rfc822.3
.\"  <!-- $Id: rfc822.3,v 1.1.1.1 2003/05/07 02:14:05 lfan Exp $ -->
.\"  <!-- Copyright 2001 Double Precision, Inc.  See COPYING for -->
.\"  <!-- distribution information. -->
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "RFC822" "3" "12 December 2002" "Double Precision, Inc." ""
.SH NAME
rfc822 \- RFC 822 parsing library
.SH SYNOPSIS
.PP

.nf
#include <rfc822.h>

#include <rfc2047.h>

cc ... -lrfc822
.fi
.SH "DESCRIPTION"
.PP
The rfc822 library provides functions for parsing E-mail headers in the RFC
822 format. This library also includes some functions to help with encoding
and decoding 8-bit text, as defined by RFC 2047.
.PP
The format used by E-mail headers to encode sender and recipient
information is defined by
RFC 822
(and its successor,
RFC 2822).
The format allows the actual E-mail
address and the sender/recipient name to be expressed together, for example:
John Smith <jsmith@example.com>
.PP
The main purposes of the rfc822 library is to:
.PP
1) Parse a text string containing a list of RFC 822-formatted address into
its logical components: names and E-mail addresses.
.PP
2) Access those individual components.
.PP
3) Allow some limited modifications of the parsed structure, and then
convert it back into a text string.
.SS "TOKENIZING AN E-MAIL HEADER"
.PP

.nf
struct rfc822t *tokens=rfc822t_alloc_new(const char *header,
                void (*err_func)(const char *, int, void *),
                void *func_arg);

void rfc822t_free(tokens);
.fi
.PP
The \fBrfc822t_alloc_new\fR() function (superceeds
\fBrfc822t_alloc\fR(), which is now
obsolete) accepts an E-mail \fIheader\fR, and parses it into
individual tokens. This function allocates and returns a pointer to an
rfc822t
structure, which is later used by
\fBrfc822a_alloc\fR() to extract
individual addresses from these tokens.
.PP
If \fIerr_func\fR argument, if not NULL, is a pointer
to a callback
function.  The function is called in the event that the E-mail header is
corrupted to the point that it cannot even be parsed.  This is a rare instance
-- most forms of corruption are still valid at least on the lexical level. 
The only time this error is reported is in the event of mismatched
parenthesis, angle brackets, or quotes.  The callback function receives the
\fIheader\fR pointer, an index to the syntax error in the
header string, and the \fIfunc_arg\fR argument.
.PP
The semantics of \fIerr_func\fR are subject to change.  It is recommended
to leave this argument as NULL in the current version of the library.
.PP
\fBrfc822t_alloc\fR() returns a pointer to a
dynamically-allocated rfc822t
structure. A NULL pointer is returned if there's insufficient memory to
allocate this structure. The \fBrfc822t_free\fR() function
destroys
rfc822t structure and frees all
dynamically allocated memory.
.sp
.RS
.B "Note:"
Until \fBrfc822t_free\fR() is called, the contents of
\fIheader\fR MUST
NOT be destroyed or altered in any way. The contents of
\fIheader\fR are not
modified by \fBrfc822t_alloc\fR(), however the
rfc822t structure contains
pointers to portions of the supplied \fIheader\fR,
and they must remain valid.
.RE
.SS "EXTRACTING E-MAIL ADDRESSES"
.PP

.nf
struct rfc822a *addrs=rfc822a_alloc(struct rfc822t *tokens);

void rfc822a_free(addrs);
.fi
.PP
The \fBrfc822a_alloc\fR() function returns a
dynamically-allocated rfc822a
structure, that contains individual addresses that were logically parsed
from a rfc822t structure.  The
\fBrfc822a_alloc\fR() function returns NULL if
there was insufficient memory to allocate the rfc822a structure. The
\fBrfc822a_free\fR() function destroys the rfc822a function, and frees all
associated dynamically-allocated memory. The rfc822t structure passed
to \fBrfc822a_alloc\fR() must not be destroyed before \fBrfc822a_free\fR() destroys the
rfc822a structure.
.PP
The rfc822a structure has the following fields:
.PP

.nf
struct rfc822a {
        struct rfc822addr *addrs;
        int     naddrs;
} ;
.fi
.PP
The \fBnaddrs\fR field gives the number of
rfc822addr structures
that are pointed to by \fBaddrs\fR, which is an array.
Each rfc822addr
structure represents either an address found in the original E-mail header,
\fBor the contents of some legacy "syntactical sugar"\fR.
For example, the
following is a valid E-mail header:
.PP

.nf
To: recipient-list: tom@example.com, john@example.com;
.fi
.PP
Typically, all of this, except for "To:",
is tokenized by \fBrfc822t_alloc\fR(), then parsed by
\fBrfc822a_alloc\fR().
"recipient-list:" and
the trailing semicolon is a legacy mailing list specification that is no
longer in widespread use, but must still must be accounted for. The resulting
rfc822a structure will have four
rfc822addr structures: one for
"recipient-list:";
one for each address; and one for the trailing semicolon.
Each rfc822a structure has the following
fields:
.PP

.nf
struct rfc822addr {
        struct rfc822token *tokens;
        struct rfc822token *name;
} ;
.fi
.PP
If \fBtokens\fR is a null pointer, this structure
represents some
non-address portion of the original header, such as
"recipient-list:" or a
semicolon.  Otherwise it points to a structure that represents the E-mail
address in tokenized form.
.PP
\fBname\fR either points to the tokenized form of a
non-address portion of
the original header, or to a tokenized form of the recipient's name.
\fBname\fR will be NULL if the recipient name was not provided. For the
following address:
Tom Jones <tjones@example.com> - the
\fBtokens\fR field points to the tokenized form of
"tjones@example.com",
and \fBname\fR points to the tokenized form of
"Tom Jones".
.PP
Each rfc822token structure contains the following
fields:
.PP

.nf
struct rfc822token {
        struct rfc822token *next;
        int token;
        const char *ptr;
        int len;
} ;
.fi
.PP
The \fBnext\fR pointer builds a linked list of all
tokens in this name or
address.  The possible values for the \fBtoken\fR field
are:
.TP
\fB0x00\fR
This is a simple atom - a sequence of non-special characters that
is delimited by whitespace or special characters (see below).
.TP
\fB0x22\fR
The value of the ascii quote - this is a quoted string.
.TP
\fBOpen parenthesis: '('\fR
This is an old style comment.  A deprecated form of E-mail
addressing uses - for example -
"john@example.com (John Smith)" instead of
"John Smith <john@example.com>".
This old-style notation defined
parenthesized content as arbitrary comments.
The rfc822token with
\fBtoken\fR set to '(' is created for the contents of
the entire comment.
.TP
\fBSymbols: '<', '>', '@', and many others\fR
The remaining possible values of \fBtoken\fR include all
the characters in RFC 822 headers that have special significance.
.PP
When a rfc822token structure does not represent a
special character, the \fBptr\fR field points to a text
string giving its contents.
The contents are NOT null-terminated, the \fBlen\fR
field contains the number of characters included.
The macro rfc822_is_atom(token) indicates whether
\fBptr\fR and \fBlen\fR are used for
the given \fBtoken\fR.
Currently \fBrfc822_is_atom\fR() returns true if
\fBtoken\fR is a zero byte, '"', or
\&'('.
.PP
Note that it's possible that \fBlen\fR might be zero.
This happens with null addresses used as return addresses for delivery status
notifications.
.SS "WORKING WITH E-MAIL ADDRESSES"
.PP

.nf
void rfc822_deladdr(struct rfc822a *addrs, int index);

void rfc822tok_print(const struct rfc822token *list,
        void (*func)(char, void *), void *func_arg);

void rfc822_print(const struct rfc822a *addrs,
        void (*print_func)(char, void *),
        void (*print_separator)(const char *, void *), void *callback_arg);
 
void rfc822_addrlist(const struct rfc822a *addrs,
                void (*print_func)(char, void *),
                void *callback_arg);
 
void rfc822_namelist(const struct rfc822a *addrs,
                void (*print_func)(char, void *),
                void *callback_arg);

void rfc822_praddr(const struct rfc822a *addrs,
                int index,
                void (*print_func)(char, void *),
                void *callback_arg);

void rfc822_prname(const struct rfc822a *addrs,
                int index,
                void (*print_func)(char, void *),
                void *callback_arg);

void rfc822_prname_orlist(const struct rfc822a *addrs,
                int index,
                void (*print_func)(char, void *),
                void *callback_arg);

char *rfc822_gettok(const struct rfc822token *list);
char *rfc822_getaddrs(const struct rfc822a *addrs);
char *rfc822_getaddr(const struct rfc822a *addrs, int index);
char *rfc822_getname(const struct rfc822a *addrs, int index);
char *rfc822_getname_orlist(const struct rfc822a *addrs, int index);

char *rfc822_getaddrs_wrap(const struct rfc822a *, int);
.fi
.PP
These functions are used to work with individual addresses that are parsed
by \fBrfc822a_alloc\fR().
.PP
\fBrfc822_deladdr\fR() removes a single
rfc822addr structure, whose
\fIindex\fR is given, from the address array in
rfc822addr.
\fBnaddrs\fR is decremented by one.
.PP
\fBrfc822tok_print\fR() converts a tokenized
\fIlist\fR of rfc822token
objects into a text string. The callback function,
\fIfunc\fR, is called one
character at a time, for every character in the tokenized objects. An
arbitrary pointer, \fIfunc_arg\fR, is passed unchanged as
the additional argument to the callback function.
\fBrfc822tok_print\fR() is not usually the most
convenient and efficient function, but it has its uses.
.PP
\fBrfc822_print\fR() takes an entire
rfc822a structure, and uses the
callback functions to print the contained addresses, in their original form,
separated by commas. The function pointed to by
\fIprint_func\fR is used to
print each individual address, one character at a time.  Between the
addresses, the \fIprint_separator\fR function is called to
print the address separator, usually the string ", ".
The \fIcallback_arg\fR argument is passed
along unchanged, as an additional argument to these functions.
.PP
The functions \fBrfc822_addrlist\fR() and
\fBrfc822_namelist\fR() also print the
contents of the entire rfc822a structure, but in a
different way.
\fBrfc822_addrlist\fR() prints just the actual E-mail
addresses, not the recipient
names or comments.  Each E-mail address is followed by a newline character.
\fBrfc822_namelist\fR() prints just the names or comments,
followed by newlines.
.PP
The functions \fBrfc822_praddr\fR() and
\fBrfc822_prname\fR() are just like
\fBrfc822_addrlist\fR() and
\fBrfc822_namelist\fR(), except that they print a single name
or address in the rfc822a structure, given its
\fIindex\fR. The
functions \fBrfc822_gettok\fR(),
\fBrfc822_getaddrs\fR(), \fBrfc822_getaddr\fR(),
and \fBrfc822_getname\fR() are equivalent to
\fBrfc822tok_print\fR(), \fBrfc822_print\fR(),
\fBrfc822_praddr\fR() and \fBrfc822_prname\fR(),
but, instead of using a callback function
pointer, these functions write the output into a dynamically allocated buffer.
That buffer must be destroyed by \fBfree\fR(3) after use.
These functions will
return a null pointer in the event of a failure to allocate memory for the
buffer.
.PP
\fBrfc822_prname_orlist\fR() is similar to
\fBrfc822_prname\fR(), except that it will
also print the legacy RFC822 group list syntax (which are also parsed by
\fBrfc822a_alloc\fR()).  \fBrfc822_praddr\fR()
will print an empty string for an index
that corresponds to a group list name (or terminated semicolon).
\fBrfc822_prname\fR() will also print an empty string.
\fBrfc822_prname_orlist\fR() will
instead print either the name of the group list, or a single string ";".
\fBrfc822_getname_orlist\fR() will instead save it into a
dynamically allocated buffer.
.PP
The function \fBrfc822_getaddrs_wrap\fR() is similar to
\fBrfc822_getaddrs\fR(), except
that the generated text is wrapped on or about the 73rd column, using
newline characters.
.SS "WORKING WITH DATES"
.PP

.nf
time_t timestamp=rfc822_parsedt(const char *datestr)
const char *datestr=rfc822_mkdate(time_t timestamp);
void rfc822_mkdate_buf(time_t timestamp, char *buffer);
.fi
.PP
These functions convert between timestamps and dates expressed in the
Date: E-mail header format.
.PP
\fBrfc822_parsedt\fR() returns the timestamp corresponding to
the given date string (0 if there was a syntax error).
.PP
\fBrfc822_mkdate\fR() returns a date string corresponding to
the given timestamp.
\fBrfc822_mkdate_buf\fR() writes the date string into the
given buffer instead,
which must be big enough to accommodate it.
.SS "WORKING WITH 8-BIT MIME-ENCODED HEADERS"
.PP

.nf
int error=rfc2047_decode(const char *text,
                int (*callback_func)(const char *, int, const char *, void *),
                void *callback_arg);
 
extern char *str=rfc2047_decode_simple(const char *text);
 
extern char *str=rfc2047_decode_enhanced(const char *text,
                const char *charset);
 
void rfc2047_print(const struct rfc822a *a,
        const char *charset,
        void (*print_func)(char, void *),
        void (*print_separator)(const char *, void *), void *);

 
char *buffer=rfc2047_encode_str(const char *string,
                const char *charset);
 
int error=rfc2047_encode_callback(const char *string,
        const char *charset,
        int (*func)(const char *, size_t, void *),
        void *callback_arg);
 
char *buffer=rfc2047_encode_header(const struct rfc822a *a,
        const char *charset);
.fi
.PP
These functions provide additional logic to encode or decode 8-bit content
in 7-bit RFC 822 headers, as specified in RFC 2047.
.PP
\fBrfc2047_decode\fR() is a basic RFC 2047 decoding function.
It receives a
pointer to some 7bit RFC 2047-encoded text, and a callback function.  The
callback function is repeatedly called. Each time it's called it receives a
piece of decoded text. The arguments are: a pointer to a text fragment, number
of bytes in the text fragment, followed by a pointer to the character set of
the text fragment. The character set pointer is NULL for portions of the
original text that are not RFC 2047-encoded.
.PP
The callback function also receives \fIcallback_arg\fR, as
its last
argument. If the callback function returns a non-zero value,
\fBrfc2047_decode\fR()
terminates, returning that value.  Otherwise,
\fBrfc2047_decode\fR() returns 0 after
a successful decoding. \fBrfc2047_decode\fR() returns -1 if it
was unable to allocate sufficient memory.
.PP
\fBrfc2047_decode_simple\fR() and
\fBrfc2047_decode_enhanced\fR() are alternatives to
\fBrfc2047_decode\fR() which forego a callback function, and
return the decoded text
in a dynamically-allocated memory buffer. The buffer must be
\fBfree\fR(3)-ed after
use. \fBrfc2047_decode_simple\fR() discards all character set
specifications, and
merely decodes any 8-bit text. \fBrfc2047_decode_enhanced\fR()
is a compromise to
discarding all character set information.  The local character set being used
is specified as the second argument to
\fBrfc2047_decode_enhanced\fR().  Any RFC
2047-encoded text in a different character set will be prefixed by the name of
the character set, in brackets, in the resulting output.
.PP
\fBrfc2047_decode_simple\fR() and
\fBrfc2047_decode_enhanced\fR() return a null pointer
if they are unable to allocate sufficient memory.
.PP
The \fBrfc2047_print\fR() function is equivalent to
\fBrfc822_print\fR(), followed by
\fBrfc2047_decode_enhanced\fR() on the result.  The callback
functions are used in
an identical fashion, except that they receive text that's already
decoded.
.PP
The function \fBrfc2047_encode_str\fR() takes a
\fIstring\fR and \fIcharset\fR
being the name of the local character set, then encodes any 8-bit portions of
\fIstring\fR using RFC 2047 encoding.
\fBrfc2047_encode_str\fR() returns a
dynamically-allocated buffer with the result, which must be
\fBfree\fR(3)-ed after
use, or NULL if there was insufficient memory to allocate the buffer.
.PP
The function \fBrfc2047_encode_callback\fR() is similar to
\fBrfc2047_encode_str\fR()
except that the callback function is repeatedly called to received the
encoding string.  Each invocation of the callback function receives a pointer
to a portion of the encoded text, the number of characters in this portion,
and \fIcallback_arg\fR.
.PP
The function \fBrfc2047_encode_header\fR() is basically
equivalent to \fBrfc822_getaddrs\fR(), followed by
\fBrfc2047_encode_str\fR();
.SS "WORKING WITH SUBJECTS"
.PP

.nf
char *basesubj=rfc822_coresubj(const char *subj);

char *basesubj=rfc822_coresubj_nouc(const char *subj);
.fi
.PP
This function takes the contents of the subject header, and returns the
"core" subject header that's used in the specification of the IMAP THREAD
function. This function is designed to strip all subject line artifacts that
might've been added in the process of forwarding or replying to a message.
Currently, \fBrfc822_coresubj\fR() performs the following transformations:
.TP
\fBWhitespace\fR
Leading and trailing whitespace is removed.  Consecutive
whitespace characters are collapsed into a single whitespace character.
All whitespace characters are replaced by a space.
.TP
\fBRe:, (fwd) [foo]\fR
These artifacts (and several others) are removed from
the subject line.
.PP
Note that this function does NOT do MIME decoding.  In order to
implement IMAP THREAD, it is necessary to call something like
\fBrfc2047_decode\fR() before
calling \fBrfc822_coresubj\fR().
.PP
This function returns a pointer to a dynamically-allocated buffer, which
must be \fBfree\fR(3)-ed after use.
.PP
\fBrfc822_coresubj_nouc\fR() is like
\fBrfc822_coresubj\fR(), except that the subject
is not converted to uppercase.
.SH "SEE ALSO"
.PP
\fBrfc2045\fR(3),
\fBreformail\fR(1),
\fBreformime\fR(1).


src/rfc822/rfc822.c
/*
** Copyright 1998 - 2001 Double Precision, Inc.
** See COPYING for distribution information.
*/

/*
** $Id: rfc822.c,v 1.1.1.1 2003/05/07 02:14:04 lfan Exp $
*/
#include	<stdio.h>
#include	<ctype.h>
#include	<stdlib.h>
#include	<string.h>
#include	"rfc822.h"

static void tokenize(const char *p, struct rfc822token *tokp, int *toklen,
	void (*err_func)(const char *, int, void *), void *voidp)
{
const char *addr=p;
int	i=0;
int	inbracket=0;

	*toklen=0;
	while (*p)
	{
		if (isspace((int)(unsigned char)*p))
		{
			p++;
			i++;
			continue;
		}

		switch (*p)	{
		int	level;

		case '(':
			if (tokp)
			{
				tokp->token='(';
				tokp->ptr=p;
				tokp->len=0;
			}
			level=0;
			for (;;)
			{
				if (!*p)
				{
					if (err_func) (*err_func)(addr, i,
								  voidp);
					if (tokp) tokp->token='"';
					++*toklen;
					return;
				}
				if (*p == '(')
					++level;
				if (*p == ')' && --level == 0)
				{
					p++;
					i++;
					if (tokp)	tokp->len++;
					break;
				}
				if (*p == '\\' && p[1])
				{
					p++;
					i++;
					if (tokp)	tokp->len++;
				}

				i++;
				if (tokp)	tokp->len++;
				p++;
			}
			if (tokp)	++tokp;
			++*toklen;
			continue;

		case '"':
			p++;
			i++;

			if (tokp)
			{
				tokp->token='"';
				tokp->ptr=p;
			}
			while (*p != '"')
			{
				if (!*p)
				{
					if (err_func) (*err_func)(addr, i,
								  voidp);
					++*toklen;
					return;
				}
				if (*p == '\\' && p[1])
				{
					if (tokp)	tokp->len++;
					p++;
					i++;
				}
				if (tokp)	tokp->len++;
				p++;
				i++;
			}
			++*toklen;
			if (tokp)	++tokp;
			p++;
			i++;
			continue;
		case '\\':
		case ')':
			if (err_func) (*err_func)(addr, i, voidp);
			++p;
			++i;
			continue;
		case '<':
		case '>':
		case '@':
		case ',':
		case ';':
		case ':':
		case '.':
		case '[':
		case ']':
		case '%':
		case '!':
		case '?':
		case '=':
		case '/':

			if ( (*p == '<' && inbracket) ||
				(*p == '>' && !inbracket))
			{
				if (err_func) (*err_func)(addr, i, voidp);
				++p;
				++i;
				continue;
			}

			if (*p == '<')
				inbracket=1;

			if (*p == '>')
				inbracket=0;

			if (tokp)
			{
				tokp->token= *p;
				tokp->ptr=p;
				tokp->len=1;
				++tokp;
			}
			++*toklen;

			if (*p == '<' && p[1] == '>')
					/* Fake a null address */
			{
				if (tokp)
				{
					tokp->token=0;
					tokp->ptr="";
					tokp->len=0;
					++tokp;
				}
				++*toklen;
			}
			++p;
			++i;
			continue;
		default:

			if (tokp)
			{
				tokp->token=0;
				tokp->ptr=p;
				tokp->len=0;
			}
			while (*p && !isspace((int)(unsigned char)*p) && strchr(
				"<>@,;:.[]()%!\"\\?=/", *p) == 0)
			{
				if (tokp)	++tokp->len;
				++p;
				++i;
			}
			if (i == 0)	/* Idiot check */
			{
				if (err_func) (*err_func)(addr, i, voidp);
				if (tokp)
				{
					tokp->token='"';
					tokp->ptr=p;
					tokp->len=1;
					++tokp;
				}
				++*toklen;
				++p;
				++i;
				continue;
			}
			if (tokp)	++tokp;
			++*toklen;
		}
	}
}

static void parseaddr(struct rfc822token *tokens, int ntokens,
		struct rfc822addr *addrs, int *naddrs)
{
int	flag, j, k;
struct	rfc822token	save_token;

	*naddrs=0;

	while (ntokens)
	{
	int	i;

		/* atoms (token=0) or quoted strings, followed by a : token
		is a list name. */

		for (i=0; i<ntokens; i++)
			if (tokens[i].token && tokens[i].token != '"')
				break;
		if (i < ntokens && tokens[i].token == ':')
		{
			++i;
			if (addrs)
			{
				addrs->tokens=0;
				addrs->name=i ? tokens:0;
				for (j=1; j<i; j++)
					addrs->name[j-1].next=addrs->name+j;
				if (i)
					addrs->name[i-1].next=0;
				addrs++;
			}
			++*naddrs;
			tokens += i;
			ntokens -= i;
			continue;  /* Group=phrase ":" */
		}

		/* Spurious commas are skipped, ;s are recorded */

		if (tokens->token == ',' || tokens->token == ';')
		{
			if (tokens->token == ';')
			{
				if (addrs)
				{
					addrs->tokens=0;
					addrs->name=tokens;
					addrs->name->next=0;
					addrs++;
				}
				++*naddrs;
			}
			++tokens;
			--ntokens;
			continue;
		}

		/* If we can find a '<' before the next comma or semicolon,
		we have new style RFC path address */

		for (i=0; i<ntokens && tokens[i].token != ';' &&
				tokens[i].token != ',' &&
					tokens[i].token != '<'; i++)
			;

		if (i < ntokens && tokens[i].token == '<')
		{
		int	j;

			/* Ok -- what to do with the stuff before '>'???
			If it consists exclusively of atoms, leave them alone.
			Else, make them all a quoted string. */

			for (j=0; j<i && (tokens[j].token == 0 ||
					tokens[j].token == '('); j++)
				;

			if (j == i)
			{
				if (addrs)
				{
					addrs->name= i ? tokens:0;
					for (k=1; k<i; k++)
						addrs->name[k-1].next=addrs->name+k;
					if (i)
						addrs->name[i-1].next=0;
				}
			}
			else	/* Intentionally corrupt the original toks */
			{
				if (addrs)
				{
					tokens->len= tokens[i-1].ptr
							+ tokens[i-1].len
							- tokens->ptr;
					/* We know that all the ptrs point
					to parts of the same string. */
					tokens->token='"';
						/* Quoted string. */
					addrs->name=tokens;
					addrs->name->next=0;
				}
			}

		/* Any comments in the name part are changed to quotes */

			if (addrs)
			{
			struct rfc822token *t;

				for (t=addrs->name; t; t=t->next)
					if (t->token == '(')
						t->token='"';
			}

			/* Now that's done and over with, see what can
			be done with the <...> part. */

			++i;
			tokens += i;
			ntokens -= i;
			for (i=0; i<ntokens && tokens[i].token != '>'; i++)
				;
			if (addrs)
			{
				addrs->tokens=i ? tokens:0;
				for (k=1; k<i; k++)
					addrs->tokens[k-1].next=addrs->tokens+k;
				if (i)
					addrs->tokens[i-1].next=0;
				++addrs;
			}
			++*naddrs;
			tokens += i;
			ntokens -= i;
			if (ntokens)	/* Skip the '>' token */
			{
				--ntokens;
				++tokens;
			}
			continue;
		}

		/* Ok - old style address.  Assume the worst */

		/* Try to figure out where the address ends.  It ends upon:
		a comma, semicolon, or two consecutive atoms. */

		flag=0;
		for (i=0; i<ntokens && tokens[i].token != ',' &&
			tokens[i].token != ';'; i++)
		{
			if (tokens[i].token == '(')	continue;
					/* Ignore comments */
			if (tokens[i].token == 0 || tokens[i].token == '"')
								/* Atom */
			{
				if (flag)	break;
				flag=1;
			}
			else	flag=0;
		}
		if (i == 0)	/* Must be spurious comma, or something */
		{
			++tokens;
			--ntokens;
			continue;
		}

		if (addrs)
		{
			addrs->name=0;
		}

		/* Ok, now get rid of embedded comments in the address.
		Consider the last comment to be the real name */

		if (addrs)
		{

			save_token.ptr=0;
			save_token.len=0;

			for (j=k=0; j<i; j++)
			{
				if (tokens[j].token == '(')
				{
					save_token=tokens[j];
					continue;
				}
				tokens[k]=tokens[j];
				k++;
			}

			if (save_token.ptr)
			{
				tokens[i-1]=save_token;
				addrs->name=tokens+i-1;
				addrs->name->next=0;
			}
			addrs->tokens=k ? tokens:NULL;
			for (j=1; j<k; j++)
				addrs->tokens[j-1].next=addrs->tokens+j;
			if (k)
				addrs->tokens[k-1].next=0;
			++addrs;
		}
		++*naddrs;
		tokens += i;
		ntokens -= i;
	}
}

static void print_token(const struct rfc822token *token,
		void (*print_func)(char, void *), void *ptr)
{
const char *p;
int	n;

	if (token->token == 0 || token->token == '(')
	{
		for (n=token->len, p=token->ptr; n; --n, ++p)
			(*print_func)(*p, ptr);
		return;
	}

	if (token->token != '"')
	{
		(*print_func)(token->token, ptr);
		return;
	}

	(*print_func)('"', ptr);
	n=token->len;
	p=token->ptr;
	while (n)
	{
		if (*p == '"' || (*p == '\\' && n == 1)) (*print_func)('\\', ptr);
		if (*p == '\\' && n > 1)
		{
			(*print_func)('\\', ptr);
			++p;
			--n;
		}
		(*print_func)(*p++, ptr);
		--n;
	}
	(*print_func)('"', ptr);
}

void rfc822tok_print(const struct rfc822token *token,
		void (*print_func)(char, void *), void *ptr)
{
int	prev_isatom=0;
int	isatom;

	while (token)
	{
		isatom=rfc822_is_atom(token->token);
		if (prev_isatom && isatom)
			(*print_func)(' ', ptr);
		print_token(token, print_func, ptr);
		prev_isatom=isatom;
		token=token->next;
	}
}

void rfc822_print(const struct rfc822a *rfcp, void (*print_func)(char, void *),
	void (*print_separator)(const char *s, void *), void *ptr)
{
	rfc822_print_common(rfcp, 0, 0, print_func, print_separator, ptr);
}

void rfc822_print_common(const struct rfc822a *rfcp,
	char *(*decode_func)(const char *, const char *), const char *chset,
	void (*print_func)(char, void *),
	void (*print_separator)(const char *, void *), void *ptr)
{
const struct rfc822addr *addrs=rfcp->addrs;
int naddrs=rfcp->naddrs;

	while (naddrs)
	{
		if (addrs->tokens == 0)
		{
			rfc822tok_print(addrs->name, print_func, ptr);
			++addrs;
			--naddrs;
			if (addrs[-1].name && naddrs)
			{
			struct	rfc822token *t;

				for (t=addrs[-1].name; t && t->next; t=t->next)
					;

				if (t && (t->token == ':' || t->token == ';'))
					(*print_separator)(" ", ptr);
			}
			continue;
		}
		else if (addrs->name && addrs->name->token == '(')
		{	/* old style */
		char *p;

			rfc822tok_print(addrs->tokens, print_func, ptr);
			(*print_func)(' ', ptr);

			if (decode_func && (p=rfc822_gettok(addrs->name))!=0)
			{
			char *q= (*decode_func)(p, chset);
			char *r;

				for (r=q; r && *r; r++)
					(*print_func)( (int)(unsigned char)*r,
						ptr);
				if (q)	free(q);
				free(p);
			}
			else rfc822tok_print(addrs->name, print_func, ptr);
		}
		else
		{
		int	print_braces=0;
		char *p;

			if (addrs->name)
			{
				if (decode_func &&
					(p=rfc822_gettok(addrs->name)) != 0)
				{
				char *q= (*decode_func)(p, chset);
				char *r;

					for (r=q; r && *r; r++)
						(*print_func)(
							(int)(unsigned char)*r,
							ptr);
					if (q)	free(q);
					free(p);
				}
				else rfc822tok_print(addrs->name,
					print_func, ptr);
				(*print_func)(' ', ptr);
				print_braces=1;
			}
			else
			{
			struct rfc822token *p;

				for (p=addrs->tokens; p && p->next; p=p->next)
					if (rfc822_is_atom(p->token) &&
						rfc822_is_atom(p->next->token))
					print_braces=1;
			}
			if (print_braces)
				(*print_func)('<', ptr);
			rfc822tok_print(addrs->tokens, print_func, ptr);
			if (print_braces)
			{
				(*print_func)('>', ptr);
			}
		}
		++addrs;
		--naddrs;
		if (naddrs)
			if (addrs->tokens || (addrs->name &&
				rfc822_is_atom(addrs->name->token)))
				(*print_separator)(", ", ptr);
	}
}

void rfc822t_free(struct rfc822t *p)
{
	if (p->tokens)	free(p->tokens);
	free(p);
}

void rfc822a_free(struct rfc822a *p)
{
	if (p->addrs)	free(p->addrs);
	free(p);
}

void rfc822_deladdr(struct rfc822a *rfcp, int index)
{
int	i;

	if (index < 0 || index >= rfcp->naddrs)	return;

	for (i=index+1; i<rfcp->naddrs; i++)
		rfcp->addrs[i-1]=rfcp->addrs[i];
	if (--rfcp->naddrs == 0)
	{
		free(rfcp->addrs);
		rfcp->addrs=0;
	}
}

static void compat_err_func(const char *p, int i, void *voidp)
{
	void (*err_func)(const char *, int)=
		(void (*)(const char *, int))voidp;

	if (err_func)
		(*err_func)(p, i);
}

struct rfc822t *rfc822t_alloc(const char *addr,
			      void (*err_func)(const char *, int))
{
	return (rfc822t_alloc_new(addr, &compat_err_func, (void *)err_func));
}


struct rfc822t *rfc822t_alloc_new(const char *addr,
	void (*err_func)(const char *, int, void *), void *voidp)
{
struct rfc822t *p=(struct rfc822t *)malloc(sizeof(struct rfc822t));

	if (!p)	return (NULL);
	memset(p, 0, sizeof(*p));

	tokenize(addr, NULL, &p->ntokens, err_func, voidp);
	p->tokens=p->ntokens ? (struct rfc822token *)
			calloc(p->ntokens, sizeof(struct rfc822token)):0;
	if (p->ntokens && !p->tokens)
	{
		rfc822t_free(p);
		return (NULL);
	}
	tokenize(addr, p->tokens, &p->ntokens, NULL, NULL);
	return (p);
}

struct rfc822a *rfc822a_alloc(struct rfc822t *t)
{
struct rfc822a *p=(struct rfc822a *)malloc(sizeof(struct rfc822a));

	if (!p)	return (NULL);
	memset(p, 0, sizeof(*p));

	parseaddr(t->tokens, t->ntokens, NULL, &p->naddrs);
	p->addrs=p->naddrs ? (struct rfc822addr *)
			calloc(p->naddrs, sizeof(struct rfc822addr)):0;
	if (p->naddrs && !p->addrs)
	{
		rfc822a_free(p);
		return (NULL);
	}
	parseaddr(t->tokens, t->ntokens, p->addrs, &p->naddrs);
	return (p);
}

void rfc822_praddr(const struct rfc822a *rfcp, int index,
	void (*print_func)(char, void *), void *ptr)
{
const struct rfc822addr *addrs;

	if (index < 0 || index >= rfcp->naddrs)	return;

	addrs=rfcp->addrs+index;
	if (addrs->tokens)
	{
		rfc822tok_print(addrs->tokens, print_func, ptr);
		(*print_func)('\n', ptr);
	}
}

void rfc822_addrlist(const struct rfc822a *rfcp,
		void (*print_func)(char, void *), void *ptr)
{
int	i;

	for (i=0; i<rfcp->naddrs; i++)
		rfc822_praddr(rfcp, i, print_func, ptr);
}

void rfc822_prname(const struct rfc822a *rfcp, int index,
	void (*print_func)(char, void *), void *ptr)
{
const struct rfc822addr *addrs;

	if (index < 0 || index >= rfcp->naddrs)	return;

	addrs=rfcp->addrs+index;

	if (!addrs->tokens)	return;
	rfc822_prname_orlist(rfcp, index, print_func, ptr);
}

void rfc822_prname_orlist(const struct rfc822a *rfcp, int index,
	void (*print_func)(char, void *), void *ptr)
{
const struct rfc822addr *addrs;

	if (index < 0 || index >= rfcp->naddrs)	return;

	addrs=rfcp->addrs+index;

	if (addrs->name)
	{
	struct rfc822token *i;
	int n;
	int	prev_isatom=0;
	int	isatom=0;

		for (i=addrs->name; i; i=i->next, prev_isatom=isatom)
		{
			isatom=rfc822_is_atom(i->token);
			if (isatom && prev_isatom)
				(*print_func)(' ', ptr);

			if (i->token == '"')
			{
				for (n=0; n<i->len; n++)
				{
					if (i->ptr[n] == '\\' &&
					    n + 1 < i->len)
						++n;
					(*print_func)(i->ptr[n], ptr);
				}
				continue;
			}

			if (i->token != '(')
			{
				print_token(i, print_func, ptr);
				continue;
			}

			for (n=2; n<i->len; n++)
				(*print_func)(i->ptr[n-1], ptr);
		}
	} else
		rfc822tok_print(addrs->tokens, print_func, ptr);
	(*print_func)('\n', ptr);
}

void rfc822_namelist(const struct rfc822a *rfcp,
	void (*print_func)(char, void *), void *ptr)
{
int	i;

	for (i=0; i<rfcp->naddrs; i++)
		rfc822_prname(rfcp, i, print_func, ptr);
}
